### PR TITLE
final version

### PR DESCRIPTION
**Pull request by AlinaSadovskaya**

### CODE REVIEW

**Review by gmltA — COMMENTED** [review]
Ok, here's the deal: I can close this PR right now, but then you won't earn a recommendation to let you pass without the exam.

Or I can ask you to refactor some things in this project and then you'll have a chance to earn one.

What would you say to that?

**Comment by AlinaSadovskaya** [comment]
let me try to fix some things)

File: app/src/main/java/com/lab3/lab3_game/Activities/GameActivity.java [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+    private DatabaseReference currentMove;
+
+
+    private ValueEventListener field_1_listener;
+    private ValueEventListener field_2_listener;
+    private ValueEventListener moveChangesListener;
+    private ValueEventListener score_1_changedListener;
+    private ValueEventListener score_2_changedListener;
+    private ValueEventListener scoreView_1_listener;
+    private ValueEventListener scoreView_2_listener;
```
Comment (gmltA) [inline app/src/main/java/com/lab3/lab3_game/Activities/GameActivity.java]: Okay the main problem here is that you blend your UI and logic layers. You work with firebase directly from the activity, while the better architecture can look like this.

**Activity <-> ViewModel (<-> GameController) <-> NetworkService <-> Firebase**

Let's break each component down:
- **Activity** (or fragment) — your UI layer. It only works with views it contains, listens to updates in ViewModel and dispatches actions back to the ViewModel
- **ViewModel** — all the data (and related business logic) related to the current view. In your case, it is a view model for the game itself (or you can split it as 2 view models - for ship placement and for the game itself).
- **GameController** — optionally you can extract most of your game logic to separate service which is not connected to UI and ViewModel in any way. But you can also handle game state directly in ViewModel. Here you should have all the actions user can perform in game.
- **NetworkService** — a wrapper around your actual network implementation. It works like a repository in some way: it takes data from your app and sends it somewhere as well as takes data from the remote source and returns it back. Why do we need this class? To introduce and adaptor that hides actual connectivity implementation.

File: app/src/main/java/com/lab3/lab3_game/Activities/GameActivity.java [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+    private TextView player_2_name_field;
+    private TextView player_1_scoreView;
+    private TextView player_2_scoreView;
+
+    private TextView myTurnView;
+    private TextView waitForMyTurnView;
+
+    private  int score1 = 0;
+    private  int score2 = 0;
+    private final int winPoints = 20;
```
Comment (gmltA) [inline app/src/main/java/com/lab3/lab3_game/Activities/GameActivity.java]: Inside ViewModel you should store **GameState** object with all these values (along with opponent's references, turns and so on. **GameController** alters this state when user performs an action and then sends updates via **NetworkService** to Firebase.

**Review by gmltA — CHANGES_REQUESTED** [review]
Let's start with these and then we'll see.

**Review by gmltA — COMMENTED** [review]
Much better. Are you going to refactor this further or that's it?)

**Comment by AlinaSadovskaya** [comment]
I thought this was the final version) but if necessary, I can add viewmodels to the rest of the activity

**Comment by gmltA** [comment]
How about game state extraction and further logic separation?

**Comment by AlinaSadovskaya** [comment]
okey) I'll try to do it

**Review by gmltA — APPROVED** [review]
Okay that's enough. Good luck.

**Pull request merged**